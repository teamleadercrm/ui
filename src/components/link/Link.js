import uiUtilities from '@teamleader/ui-utilities';
import cx from 'classnames';
import PropTypes from 'prop-types';
import React, { forwardRef, useImperativeHandle, useRef } from 'react';
import Box from '../box';
import theme from './theme.css';

const Link = forwardRef(
  ({ children, className, disabled, element, inherit, inverse, selected, onMouseUp, onMouseLeave, ...others }, ref) => {
    const linkRef = useRef();
    useImperativeHandle(ref, () => linkRef.current);

    const blur = () => {
      const currentLinkRef = linkRef.current;
      if (currentLinkRef.blur) {
        currentLinkRef.blur();
      }
    };

    const handleMouseUp = (event) => {
      blur();
      onMouseUp && onMouseUp(event);
    };

    const handleMouseLeave = (event) => {
      blur();
      onMouseLeave && onMouseLeave(event);
    };

    const classNames = cx(
      uiUtilities['reset-font-smoothing'],
      theme['link'],
      {
        [theme['is-disabled']]: disabled,
        [theme['is-inherit']]: inherit,
        [theme['is-inverse']]: inverse,
        [theme['is-selected']]: selected,
      },
      className,
    );

    return (
      <Box
        element={element}
        ref={linkRef}
        className={classNames}
        data-teamleader-ui="link"
        onMouseUp={handleMouseUp}
        onMouseLeave={handleMouseLeave}
        {...others}
      >
        {children}
      </Box>
    );
  },
);

Link.propTypes = {
  /** The content to display inside the link. */
  children: PropTypes.any.isRequired,
  /** A class name for the link to give custom styles. */
  className: PropTypes.string,
  /** If true, component will be disabled. */
  disabled: PropTypes.bool,
  /** If true, the link style inherits the parent element style. */
  inherit: PropTypes.bool,
  /** If true, the underline behavior will be inverted. */
  inverse: PropTypes.bool,
  /** A custom element to be rendered */
  element: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  /** Callback function that is fired when mouse leaves the component. */
  onMouseLeave: PropTypes.func,
  /** Callback function that is fired when the mouse button is released. */
  onMouseUp: PropTypes.func,
  /** If true, component will be shown in a selected state */
  selected: PropTypes.bool,
};

Link.defaultProps = {
  className: '',
  disabled: false,
  element: 'a',
  inherit: true,
  inverse: false,
};

Link.displayName = 'Link';

export default Link;
