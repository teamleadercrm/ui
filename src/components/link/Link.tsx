import uiUtilities from '@teamleader/ui-utilities';
import cx from 'classnames';
import React, { forwardRef, ReactNode, useImperativeHandle, useRef } from 'react';
import { GenericComponent } from '../../@types/types';
import Box from '../box';
import { BoxProps } from '../box/Box';
import theme from './theme.css';

export interface LinkProps extends Omit<BoxProps, 'ref'> {
  /** The content to display inside the link. */
  children: ReactNode;
  /** A class name for the link to give custom styles. */
  className?: string;
  /** If true, component will be disabled. */
  disabled?: boolean;
  /** If true, the link style inherits the parent element style. */
  inherit?: boolean;
  /** If true, the underline behavior will be inverted. */
  inverse?: boolean;
  /** A custom element to be rendered */
  element?: React.ElementType;
  /** Callback function that is fired when mouse leaves the component. */
  onMouseLeave?: (event: React.MouseEvent) => void;
  /** Callback function that is fired when the mouse button is released. */
  onMouseUp?: (event: React.MouseEvent) => void;
  /** If true, component will be shown in a selected state */
  selected?: boolean;
}

const Link: GenericComponent<LinkProps> = forwardRef<HTMLElement, LinkProps>(
  (
    {
      children,
      className = '',
      disabled = false,
      element = 'a',
      inherit = true,
      inverse = false,
      selected,
      onMouseUp,
      onMouseLeave,
      ...others
    },
    ref,
  ) => {
    const linkRef = useRef<HTMLElement>(null);
    useImperativeHandle<HTMLElement | null, HTMLElement | null>(ref, () => linkRef.current);

    const blur = () => {
      const currentLinkRef = linkRef.current;
      if (currentLinkRef?.blur) {
        currentLinkRef.blur();
      }
    };

    const handleMouseUp = (event: React.MouseEvent) => {
      blur();
      onMouseUp && onMouseUp(event);
    };

    const handleMouseLeave = (event: React.MouseEvent) => {
      blur();
      onMouseLeave && onMouseLeave(event);
    };

    const classNames = cx(
      uiUtilities['reset-font-smoothing'],
      theme['link'],
      {
        [theme['is-disabled']]: disabled,
        [theme['is-inherit']]: inherit,
        [theme['is-inverse']]: inverse,
        [theme['is-selected']]: selected,
      },
      className,
    );

    return (
      <Box
        element={element}
        ref={linkRef}
        className={classNames}
        data-teamleader-ui="link"
        onMouseUp={handleMouseUp}
        onMouseLeave={handleMouseLeave}
        {...others}
      >
        {children}
      </Box>
    );
  },
);

Link.displayName = 'Link';

export default Link;
