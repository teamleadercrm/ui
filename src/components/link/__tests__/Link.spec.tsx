import React from 'react';
import { render } from '@testing-library/react';
import Link from '../';

describe('Component - Link', () => {
  it('renders', async () => {
    const { asFragment } = render(<Link href="https://foobar.com">Foobar</Link>);
    expect(asFragment()).toMatchSnapshot();
  });
});
