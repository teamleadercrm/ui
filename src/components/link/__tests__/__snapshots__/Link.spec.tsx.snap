// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Link renders 1`] = `
<DocumentFragment>
  <a
    class="box reset-font-smoothing link is-inherit"
    data-teamleader-ui="link"
    href="https://foobar.com"
  >
    Foobar
  </a>
</DocumentFragment>
`;
